sdk/core/core-rest-pipeline/src/nodeHttpClient.ts

```diff
@@ -210,6 +204,24 @@ class NodeHttpClient implements HttpClient {
     return new Promise<http.IncomingMessage>((resolve, reject) => {
       const req = isInsecure ? http.request(options) : https.request(options);
 
+      if (request.timeout > 0) {
+        req.on("connect", () => {
+          setTimeout(() => req.emit("timeout"), request.timeout);
+        });
+        req.on("socket", (socket) => {
+          // Node warrantees that an instance of the <net.Socket> class will be passed,
+          // unless the user specifies a socket type other than <net.Socket>.
+          // Meaning that we may see unexpected sockets in the wild.
+          // For our tests though, we know that socket will be undefined.
+          if (socket?.setTimeout) {
+            socket.setTimeout(request.timeout);
+          } else {
+            setTimeout(() => req.emit("timeout"), request.timeout);
+          }
+        });
+        req.on("timeout", () => abortController.abort());
+      }
```

Comment:
I'm not 100% sure what you're trying to do here. The `socket` event should be invoked as soon as the http agent assigns the request a socket. The socket may already be connected (assuming keep alive is enabled and the socket has been used for a previous request), which can be checked via `socket.connecting`.

It looks like when a socket is assigned, you're immediately setting the max idle timeout via `socket.setTimeout`, which takes affect after the socket is connected, but not before. Then you also listen to the `connect` event and manually trigger the `timeout` event to be emitted after the user-defined timeout.

1. I don't see the manual timeout you mention in the changelog. It looks like if the `connect` event is never emitted and the request is assigned a new socket, you'll never timeout.
2. I don't think we should ever be emitting `timeout` on the request ourselves. This is an event that signifies that the max idle timeout has been reached on the request. We should either rely on the request emitting it itself, or if we actually intend to measure something different, find another way to do so.

I would be in favor of supporting a generic 'connectionTimeoutInMs' property that only applies for node.js requests. To do that, you can check if `socket.isConnecting === true` in the `socket` event handler, and if so set a timeout to abort the request. This timeout can then be cleared when `connect` is emitted. I'm not sure how much this helps you because we still treat `timeout` as a request timeout versus a max idle timeout...

If we could change the behavior of timeout to be the max idle timeout and rely on the abort signal instead for the max request time...that would be my ideal place to be but maybe we can say connectionTimeoutInMs is applied additively to the request timeout to fix the issue you're seeing.

Reply:
> If we could change the behavior of timeout to be the max idle timeout and rely on the abort signal instead for the max request time...that would be my ideal place to be

That sounds good to me! We can do that. cc: @xirzec

Reply:
> a generic 'connectionTimeoutInMs' property that only applies for node.js requests

That sounds cool too!

Reply:
I think we'd need to go with the 'connectionTimeoutInMs applies additively' approach for the sake of backwards compatibility. If we changed the existing timeout option to say it was the max idle timeout rather than the max time to get the response, it could cause operations to take longer than they were previously.

Reply:
Yeah we should make this change additively. Also this code doesn't look like what I'd expect.

Can we do this instead?

```ts
req.setTimeout(request.connectionTimeoutInMs);
req.once("timeout", () => { 
   req.destroy();
   reject(new RestError(
       `Socket was idle for longer than configured timeout of ${request.connectionTimeoutInMs}`, 
       { code: RestError.REQUEST_SEND_ERROR, request }
   ));
});
```

Reply:
req.setTimeout doesn’t exist :/ I’m thinking it was added later or something

Reply:
if “additive” in this case means to leave the previous timeout as is, that is not possible, since the issue is that it happens too fast 🤔

Reply:
I had a call with Christopher and I think I understand better! I’m updating this PR.

Reply:
> req.setTimeout doesn’t exist :/ I’m thinking it was added later or something

https://nodejs.org/dist/latest-v14.x/docs/api/http.html#http_request_settimeout_timeout_callback

![image](https://user-images.githubusercontent.com/639216/127232866-d77e814d-f31d-46f9-ae9b-a1f574505d20.png)

Docs say it's supported since v 0.5.9

Reply:
@xirzec I’m sorry about that! I was getting some error on the editor, but I think it disappeared after I restarted the TS server.

I’ve pushed a new design after discussing different approaches with @chradek ! Please let me know what you think of the new design @xirzec